File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
class Solution {
    private int before(TreeNode node) {
        node = node.left;
        while (node.right != null) {
            node = node.right;
        }
        return node.val;
    }

    private int next(TreeNode node) {
        node = node.right;
        while (node.left != null) {
            node = node.left;
        }
        return node.val;
    }

    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;

        if (key > root.val) {
            root.right = deleteNode(root.right, key);
        } else if (key < root.val) {
            root.left = deleteNode(root.left, key);
        } else {
            if (root.left == null && root.right == null) {
                return null;
            } else if (root.right != null) {
                root.val = next(root);
                root.right = deleteNode(root.right, root.val);
            } else {
                root.val = before(root);
                root.left = deleteNode(root.left, root.val);
            }
        }

        return root;
    }
}

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
class Solution:
    def before(self, root):
        root = root.left
        while root.right:
            root = root.right
        return root.val


    def next(self, root):
        root = root.right
        while root.left:
            root = root.left
        return root.val


    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return

        if key > root.val:
            root.right = self.deleteNode(root.right, key)
        elif key < root.val:
            root.left = self.deleteNode(root.left, key)
        else:
            if not root.left and not root.right:
                root = None
            elif root.right:
                root.val = self.next(root)
                root.right = self.deleteNode(root.right, root.val)
            else:
                root.val = self.before(root)
                root.left = self.deleteNode(root.left, root.val)
        return root